CSPLemmas/Lemmas/Lemma15.cs

```diff
@@ -0,0 +1,297 @@
+﻿using CSP;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace CSPLemmas
+{
+    public static partial class CSPLemmas
+    {
+        public static List<CspInstance> Lemma15(CspInstance instance)
+        {
+            List<CspInstance> result = new();
+            HashSet<Pair> set = findSmallBadThreeComponent(instance.Restrictions);
+            if(set != null)
+            {
+                Debug.WriteLine(set.Count);
+                if(set.Count == 12)
+                {
+                    List<Pair> list = bruteColor(set, instance.Restrictions);
+                    if(list != null)
+                    {
+                        foreach (Pair p in list)
+                        {
+                            instance.AddToResult(p);
+                        }
+                        result.AddRange(Lemma15(instance));
+                    } 
+                    else
+                    {
+                        return null;
+                    }
+                }
+                else if (set.Count == 8)
+                {
+                    List<Pair> setWithTriangles = getWithTriangle(set, instance.Restrictions);
+                    if(setWithTriangles != null) //v1, v2, w1, w2, x1, x2, y1, y2
+                    {
+                        CspInstance instance2 = instance.Clone();
+                        CspInstance instance3 = instance.Clone();
+
+                        instance.AddToResult(setWithTriangles[0]);
+                        instance.AddToResult(setWithTriangles[2]);
+                        instance.AddToResult(setWithTriangles[5]);
+
+                        instance2.AddToResult(setWithTriangles[0]);
+                        instance2.AddToResult(setWithTriangles[2]);
+                        instance2.AddToResult(setWithTriangles[7]);
+
+                        instance3.AddToResult(setWithTriangles[5]);
+                        instance3.AddToResult(setWithTriangles[7]);
+
+                        result.AddRange(Lemma15(instance));
+                        result.AddRange(Lemma15(instance2));
+                        result.AddRange(Lemma15(instance3));
```

Comment:
Testy się psują bo dodajesz do rezultatu instance2 parę która do niej nie należy.
setWithTriangles[0...7] to pary należące do instancji pierwszej. Sklonowana instancja ma inne zmienne i kolory.
Musisz użyć funkcji klonującej od Michała CloneAndReturnCorresponding(), żeby móc dodać odpowiednią parę która odpowiada parze wejściowej